# no-direct-set-state-in-use-effect

## Rule category

Correctness.

## What it does

Disallow direct calls to the [`set` function](https://react.dev/reference/react/useState#setstate) of `useState` in `useEffect`.

## Why is this bad?

Calling `setState` directly in `useEffect` can lead to infinite update loops and other side effects in an imperceptible way.

## Known limitations

- The `set` call to `useState` in the `cleanup` function of `useEffect` will not be checked.
- The current implementation does not support determining whether a `set` function called in an `async` function is actually at least one `await` after.

The limitation may be lifted in the future.

## Examples

### Failing

```tsx
import { useEffect, useState } from 'react';

function Form() {
  const [firstName, setFirstName] = useState('Taylor');
  const [lastName, setLastName] = useState('Swift');

  // ðŸ”´ Avoid: redundant state and unnecessary Effect
  const [fullName, setFullName] = useState('');
  useEffect(() => {
    setFullName(firstName + ' ' + lastName);
  }, [firstName, lastName]);
  // ...
}
```

### Passing

```tsx
import { useState } from 'react';

function Form() {
  const [firstName, setFirstName] = useState('Taylor');
  const [lastName, setLastName] = useState('Swift');
  // âœ… Good: calculated during rendering
  const fullName = firstName + ' ' + lastName;
  // ...
}
```

### Failing

```tsx
import { useEffect, useState } from 'react';

function TodoList({ todos, filter }) {
  const [newTodo, setNewTodo] = useState('');

  // ðŸ”´ Avoid: redundant state and unnecessary Effect
  const [visibleTodos, setVisibleTodos] = useState([]);
  useEffect(() => {
    setVisibleTodos(getFilteredTodos(todos, filter));
  }, [todos, filter]);

  // ...
}
```

### Passing

```tsx
import { useMemo, useState } from 'react';

function TodoList({ todos, filter }) {
  const [newTodo, setNewTodo] = useState('');
  // âœ… Does not re-run getFilteredTodos() unless todos or filter change
  const visibleTodos = useMemo(() => getFilteredTodos(todos, filter), [todos, filter]);
  // ...
}
```

### Failing

```tsx
import { useEffect, useState } from 'react';

export default function ProfilePage({ userId }) {
  const [comment, setComment] = useState('');

  // ðŸ”´ Avoid: Resetting state on prop change in an Effect
  useEffect(() => {
    setComment('');
  }, [userId]);
  // ...
}
```

### Passing

```tsx
import { useState } from 'react';

export default function ProfilePage({ userId }) {
  return (
    <Profile
      userId={userId}
      key={userId}
    />
  );
}

function Profile({ userId }) {
  // âœ… This and any other state below will reset on key change automatically
  const [comment, setComment] = useState('');
  // ...
}
```

### Failing

```tsx
import { useEffect, useState } from 'react';

function List({ items }) {
  const [isReverse, setIsReverse] = useState(false);
  const [selection, setSelection] = useState(null);

  // ðŸ”´ Avoid: Adjusting state on prop change in an Effect
  useEffect(() => {
    setSelection(null);
  }, [items]);
  // ...
}
```

### Passing

```tsx
import { useState } from 'react';

function List({ items }) {
  const [isReverse, setIsReverse] = useState(false);
  const [selection, setSelection] = useState(null);

  // Better: Adjust the state while rendering
  const [prevItems, setPrevItems] = useState(items);
  if (items !== prevItems) {
    setPrevItems(items);
    setSelection(null);
  }
  // ...
}
```

```tsx
import { useState } from 'react';

function List({ items }) {
  const [isReverse, setIsReverse] = useState(false);
  const [selectedId, setSelectedId] = useState(null);
  // âœ… Best: Calculate everything during rendering
  const selection = items.find(item => item.id === selectedId) ?? null;
  // ...
}
```

### Further Reading

- [React: useState](https://react.dev/reference/react/useState#setstate)
- [React: useEffect](https://react.dev/reference/react/useEffect)
- [React: You Might Not Need an Effect](https://react.dev/learn/you-might-not-need-an-effect)
